provider/lpwindow/submit_task.go

```diff
@@ -0,0 +1,245 @@
+	}
+
+	if head.Height() > submitByEpoch {
+		// we missed the deadline, no point in submitting
+		log.Errorw("missed submit deadline", "spID", spID, "deadline", deadline, "partition", partition, "submitByEpoch", submitByEpoch, "headHeight", head.Height())
+		return true, nil
+	}
+
+	if head.Height() < submitAtEpoch {
+		log.Errorw("submit epoch not reached", "spID", spID, "deadline", deadline, "partition", partition, "submitAtEpoch", submitAtEpoch, "headHeight", head.Height())
+		return false, xerrors.Errorf("submit epoch not reached: %d < %d", head.Height(), submitAtEpoch)
+	}
+
+	var params miner.SubmitWindowedPoStParams
+	if err := params.UnmarshalCBOR(bytes.NewReader(proofParamBytes)); err != nil {
+		return false, xerrors.Errorf("unmarshaling proof message: %w", err)
+	}
+
+	maddr, err := address.NewIDAddress(spID)
+	if err != nil {
+		return false, xerrors.Errorf("invalid miner address: %w", err)
+	}
+
+	mi, err := w.api.StateMinerInfo(context.Background(), maddr, types.EmptyTSK)
+	if err != nil {
+		return false, xerrors.Errorf("error getting miner info: %w", err)
+	}
+
+	msg := &types.Message{
+		To:     maddr,
+		Method: builtin.MethodsMiner.SubmitWindowedPoSt,
+		Params: proofParamBytes,
+		Value:  big.Zero(),
+
+		From: mi.Worker, // set worker for now for gas estimation
+	}
+
+	// calculate a more frugal estimation; premium is estimated to guarantee
+	// inclusion within 5 tipsets, and fee cap is estimated for inclusion
+	// within 4 tipsets.
+	minGasFeeMsg := *msg
+
+	minGasFeeMsg.GasPremium, err = w.api.GasEstimateGasPremium(context.Background(), 5, msg.From, msg.GasLimit, types.EmptyTSK)
+	if err != nil {
+		log.Errorf("failed to estimate minimum gas premium: %+v", err)
+		minGasFeeMsg.GasPremium = msg.GasPremium
+	}
+
+	minGasFeeMsg.GasFeeCap, err = w.api.GasEstimateFeeCap(context.Background(), &minGasFeeMsg, 4, types.EmptyTSK)
+	if err != nil {
+		log.Errorf("failed to estimate minimum gas fee cap: %+v", err)
+		minGasFeeMsg.GasFeeCap = msg.GasFeeCap
+	}
+
+	// goodFunds = funds needed for optimal inclusion probability.
+	// minFunds  = funds needed for more speculative inclusion probability.
+	goodFunds := big.Add(msg.RequiredFunds(), msg.Value)
+	minFunds := big.Min(big.Add(minGasFeeMsg.RequiredFunds(), minGasFeeMsg.Value), goodFunds)
+
+	from, _, err := w.as.AddressFor(context.Background(), w.api, mi, api.PoStAddr, goodFunds, minFunds)
+	if err != nil {
+		return false, xerrors.Errorf("error getting address: %w", err)
+	}
+
+	msg.From = from
+
+	mss := &api.MessageSendSpec{
+		MaxFee: abi.TokenAmount(w.maxWindowPoStGasFee),
+	}
+
+	smsg, err := w.sender.Send(context.Background(), msg, mss, "wdpost")
+	if err != nil {
+		return false, xerrors.Errorf("sending proof message: %w", err)
+	}
+
+	// set message_cid in the wdpost_proofs entry
+
+	_, err = w.db.Exec(context.Background(), `update wdpost_proofs set message_cid = $1 where sp_id = $2 and deadline = $3 and partition = $4`, smsg.String(), spID, deadline, partition)
```

Comment:
nit: capitalize SQL keywords for readability
UPDATE wdpost_proofs SET message_cid = $1 WHERE sp_id = $2 AND deadline = $3 AND partition = $4